xde/src/xde.rs

```diff
@@ -441,55 +441,49 @@ fn create_xde(req: &CreateXdeReq) -> Result<NoResp, OpteError> {
         None => (),
     };
 
+    let cfg = &req.cfg;
     match devs.iter().find(|x| {
-        x.vni == req.vpc_vni && x.port.mac_addr() == req.private_mac.into()
+        x.vni == cfg.vni && x.port.mac_addr() == cfg.private_mac.into()
     }) {
         Some(_) => {
             return Err(OpteError::MacExists {
                 port: req.xde_devname.clone(),
-                vni: req.vpc_vni,
-                mac: req.private_mac,
+                vni: cfg.vni,
+                mac: cfg.private_mac,
             })
         }
         None => (),
     }
 
-    // XXX-EXT-IP This is for the external IP hack.
-    let phys_gw_mac = if unsafe { xde_ext_ip_hack == 1 } && req.snat.is_some() {
-        Some(req.snat.as_ref().unwrap().phys_gw_mac)
-    } else {
-        None
-    };
-
+    // XXX-EXT-IP Copy the configuration, modifying the external IP hack
+    // fields depending on the value of the `xde_ext_ip_hack` tunable.
+    let proxy_arp_enable = unsafe { xde_ext_ip_hack == 1 };
     let vpc_cfg = VpcCfg {
-        vpc_subnet: req.vpc_subnet,
-        private_mac: req.private_mac,
-        private_ip: req.private_ip,
-        gw_mac: req.gw_mac,
-        gw_ip: req.gw_ip,
-        external_ips_v4: req.external_ips_v4,
-        snat: req.snat.clone(),
-        vni: req.vpc_vni,
-        phys_ip: req.src_underlay_addr,
-        bsvc_addr: PhysNet {
-            ether: MacAddr::from([0; 6]), //XXX this should not be needed
-            ip: req.bsvc_addr,
-            vni: req.bsvc_vni,
-        },
-        proxy_arp_enable: unsafe { xde_ext_ip_hack == 1 },
-        phys_gw_mac,
+        proxy_arp_enable,
+        phys_gw_mac: if proxy_arp_enable { cfg.phys_gw_mac } else { None },
+        ..cfg.clone()
     };
 
     // If this is the first guest in this VPC, then create a new
     // mapping for said VPC. Otherwise, pull the existing one.
-    let port_v2p = state.vpc_map.add(
-        IpAddr::Ip4(req.private_ip),
-        PhysNet {
-            ether: req.private_mac,
-            ip: req.src_underlay_addr,
-            vni: req.vpc_vni,
-        },
-    );
+    //
+    // We need to insert mappings for both IPv4 and IPv6 addresses, should the
+    // guest have them. They should return the same `Virt2Phys` mapping, since
+    // they're mapping both IP addresses to the same host.
```

Comment:
What I meant was that the mapping for a guest IPv4 and IPv6 address should return the same `Virt2Phys` object, which refers to the IPv6 address of just one host. I.e., both the IPv4 and IPv6 addresses "live" on the same host. Does that make sense? Or am I missing something? If it does, I'll add some more detail since it's clearly confusing.

Reply:
Oh duh. Sorry I just had a brain fart.